Track cloaked elements so they can be hidden again before caching

jduff opened the pull request:
This PR is working towards fixing #9 

This PR has a few distinct changes to help solve the flashing elements issue:
1. Switched to required the adapter to be loaded before Alpine. This is so we can hook into `deferLoadingAlpine` and do some work (tagging `x-cloak` elements) before Alpine starts.
2. Tag any elements with `x-cloak` on page load or turbolinks navigation so the `x-cloak` attribute can be restored when the dom is cached.

### Outstanding issues:
1. I don't like that I'm using `turbolinks:load` event to determine that the DOM is loaded and I can tag the `x-cloak` elements. I especially don't like that Alpine is initialized at the end of that event handler. It might be fine, but it feels wrong to me.
2. The tests still aren't great because they are only testing that the elements are being tagged and not testing for the outcome. There is no real test that the tagged elements get cloaked, and then hidden, correctly because those tests were passing before any of this code was written.

### Previous outdated PR description:

I don't think this PR is ready to merge, but opening it so we can continue discussing the approach. There are a few issues with the current implementation:
1. I added a couple of tests looking for the custom attribute that gets added but ideally we would test for outcomes (element was visible etc). The problem is the current code reinitializes the components so if you show or hide an element, navigate forward and back, Alpine will reinitialize the component into the correct state before the test runs.
2. ~The test for initial page load doesn't work, even when just checking for the attribute being added. I tested against a real project and it seemed to work correctly there so might be a timing issue in the test (I'm not super familiar with cypress and js testing).~ Test for initial page load is working with the latest changes.
3. The script needs to be included with `defer`, it will most likely error if it isn't included with `defer`.

I attempted to use `deferLoadingAlpine` as we discussed so it would work without the script being deferred, the problem is if Alpine is required first it will run the check the defer loading function before the bridge script executes https://github.com/alpinejs/alpine/blob/master/src/index.js#L126, so by the time the bridge script runs Alpine is already started. For this to work `alpine-turbolins-adapter` would need to be loaded before Alpine itself. I wasn't sure if this was the correct direction to take or if there was another option I might be missing.

SimoTod commented:
Hi John, thanks for your work on this PR. I'm happy to ask people to include the library before Alpine and use deferLoadingAlpine. At the moment, I didn't enforce it because it wasn't necessary but it's a common pattern for these Alpine plugins.
Could you implement it using deferLoadingAlpine and possibly update the documentation please?
I saw you added 2 event listeners as well but we are already listening for the same events above. I'd prefer to add the relevant code in the existing listener if you don't mind.
I'll try to have a look at the tests later or tomorrow and see what we can do.

jduff commented on src/bridge.js:
```diff
@@ -1,8 +1,22 @@
+
+      document.addEventListener('turbolinks:load', () => {
+        // Tag all cloaked elements on first page load.
+        this.tagCloakedElements(document.body)
```
Tagging the cloaked elements needs to happen after the dom is ready, I just used `turbolinks:load` but we could use an async await for dom ready too (like Alpine has in utils).

Alpine itself needs to be initialized after we've captured the cloaked elements, in the test (which was correctly failing before and passes now) Alpine will initialize components (which strips the `x-cloak` attribute as soon as it starts, which can happen before any other events are fired.

I feel like there is still a better way to do this where we can listen for some DOM ready event to do the tagging, and still initialize Alpine outside of that handler but I'm not seeing it yet.

jduff commented on src/bridge.js:
```diff
@@ -12,10 +26,44 @@ export default class Bridge {
+    })
+  }
+
+  removeAlpineGeneratedElements (node) {
```
These two functions just wrap up the previous event handlers, pulled them into functions so it's a bit more obvious that a few unrelated things are happening in the event handlers.

SimoTod replied:
Cool, I might unify them. I know they are unrelated but querySelectorAll is a slow operation (it gets more obvious when your page is really big) so calling it multiple times has a price. I'll sleep on it and decide tomorrow/over the weekend. PR looks good, thanks for your help.

jduff commented:
I rebuilt this on top of the latest master with a few of the suggested changes:
- Unified the `querySelectorAll` usage in each callback
- Added constants for some of the repeated attributes and the query selector strings
- removed `dist` changes from the PR

SimoTod commented:
Thanks for your contribution
